core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/OptimizeJoinsBelowEmptyAggregationRule.java

```diff
@@ -0,0 +1,156 @@
+        return aggregation.getSource().accept(new Rewriter(metadata, context.getLookup()), null)
+                .map(rewrittenSource -> Result.ofPlanNode(aggregation.replaceChildren(ImmutableList.of(rewrittenSource))))
+                .orElse(Result.empty());
+    }
+
+    private static class Rewriter
+            extends PlanVisitor<Optional<PlanNode>, Void>
+    {
+        private final Metadata metadata;
+        private final Lookup lookup;
+
+        private Rewriter(Metadata metadata, Lookup lookup)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected Optional<PlanNode> visitPlan(PlanNode node, Void context)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<PlanNode> visitFilter(FilterNode node, Void context)
+        {
+            if (!isDeterministic(node.getPredicate(), metadata)) {
+                // non-deterministic expressions could filter duplicate rows probabilistically
+                return Optional.empty();
+            }
+
+            return node.getSource().accept(this, null)
+                    .map(source -> node.replaceChildren(ImmutableList.of(source)));
+        }
+
+        @Override
+        public Optional<PlanNode> visitProject(ProjectNode node, Void context)
+        {
+            boolean isDeterministic = node.getAssignments().getExpressions().stream()
+                    .allMatch(expression -> isDeterministic(expression, metadata));
+            if (!isDeterministic) {
+                // non-deterministic projections could be used in downstream filters which could
+                // filter duplicate rows probabilistically
+                return Optional.empty();
+            }
+
+            return node.getSource().accept(this, null)
+                    .map(source -> node.replaceChildren(ImmutableList.of(source)));
+        }
+
+        @Override
+        public Optional<PlanNode> visitJoin(JoinNode node, Void context)
+        {
+            if (!isDeterministic(node.getFilter().orElse(TRUE_LITERAL), metadata)) {
+                // LookupJoinOperator will evaluate non-deterministic condition on output rows until one of the
+                // rows matches. Therefore it's safe to set maySkipOutputDuplicates for joins with non-deterministic
+                // filters.
+                if (node.isMaySkipOutputDuplicates()) {
+                    return Optional.empty();
+                }
+
+                return Optional.of(node.withMaySkipOutputDuplicates());
+            }
+
+            Optional<PlanNode> rewrittenLeft = node.getLeft().accept(this, null);
+            Optional<PlanNode> rewrittenRight = node.getRight().accept(this, null);
+
+            if (node.isMaySkipOutputDuplicates() && rewrittenLeft.isEmpty() && rewrittenRight.isEmpty()) {
+                return Optional.empty();
```

Comment:
Why would we undo `isMaySkipOutputDuplicates` for a node that has already been determined as able to do it?